u_parrots.go

```diff
@@ -938,3 +941,261 @@ func removeRC4Ciphers(s []uint16) []uint16 {
 	}
 	return s[:sliceLen]
 }
+
+// FingerprintClientHello returns a ClientHelloSpec which is based on the ClientHello that is passed in as the data argument
```

Comment:
I think the documentation should also specify whether the record header should be present.  It looks like the answer is no.  Personally, I would say it makes the most sense to accept full records, but I think that's a subjective design decision.  Either way, it's important IMO to have that specified here.

Reply:
Really good point! (on both comments)

Reply:
I stuck with accepting the Handshake Message only, just because that's what `crypto/tls.clientHelloMsg.marshal()` outputs:
https://golang.org/src/crypto/tls/handshake_messages.go

I'm open to maybe additionally adding something like handling full tls records if you think that'd be useful?

I did go ahead and make the documentation more explicit.

Reply:
I think it is more likely that someone has a full record than just the handshake data.  OTOH, if someone had only the handshake data and we only accepted full records, they'd have to create their own header which is awkward.  Starting with the logic [here](https://github.com/getlantern/tlsutil/blob/7e08a8c42b7b063b83eecca03b84e1fa4086e9e4/hello.go#L22) may give you an idea of what's required to parse the header and read the data out of a record.  I'm sure you can simplify it quite a bit if you know that you're dealing with a client hello though.

Reply:
Yeah I think if you're confident that you've got a ClientHello record, you can just strip the first 5 bytes off.  So accepting just the record data is probably fine.

Reply:
@hwh33 thought you should see that Sergey agreed w/ you :grin: https://github.com/refraction-networking/utls/pull/58/files#r472576678

Reply:
Heh, thanks 😛

That's great to see that Sergey's already gone through the PR.  Looks like he's provided some good feedback as well.

Reply:
Yeah I'm amazed he was as thorough as he was as quickly as he was considering I think his thesis/dissertation defense is tomorrow???

Reply:
Yeah, no kidding!!